refactor(es6): migrate structure to es6 classes

MichaelSolati opened the pull request:
Turned each object into a class, as the passing of functions made it difficult to neatly have the JSDocs and param types appear in intellisense (coming in another PR).

joshuaslate commented on src/create-client.js:
```diff
@@ -35,7 +35,7 @@ export const clientConstructor = ({ token, ...options }) => {
       })
         .then(response => response.json())
-        .catch(error => { throw new Error(error) })
+        .catch(error => { throw new Error(error) });
```
If we're not doing anything here other than re-throwing, we could just skip catching here.

jeremenichelli commented on src/images.js:
```diff
@@ -1,81 +1,81 @@
+      if (choiceImageSizes.includes(choiceSize)) {
+        requestQuery.url += `/choice/${choiceSize}`;
+      } else {
+        throw new Error(`Image choice size doesn't exists`);
```
typo, should be _Image choice size doesn't exist._

jeremenichelli commented on src/images.js:
```diff
@@ -1,81 +1,81 @@
-      file_name: fileName,
-      media_type: mediaType
+    if (choiceSize !== undefined) {
+      if (choiceImageSizes.includes(choiceSize)) {
```
We could join this condition here as:

```js
if (typeof choiceSize !== 'undefined' && choiceImageSizes.includes(choiceSize)) { 
  // ... 
} else {
  throw Error()
}
```

MichaelSolati replied:
Then we do an elseif for `typeof choiceSize !== 'undefined' && !choiceImageSizes.includes(choiceSize)`?

jeremenichelli replied:
I would just do else since you want to throw the error anyways. Is there a situation where the user doesn't provide a valid image size that you don't want to throw?

MichaelSolati replied:
So the `get` method was designed to fetch an image by its id as well as fetch a specific image by a size. So ideally a user can call the method without giving it a size. Aka `[https://api.typeform.com/images/{image_id}
](https://developer.typeform.com/create/reference/retrieve-image/)` AND `[https://api.typeform.com/images/{image_id}/image/{size}](https://developer.typeform.com/create/reference/retrieve-image-by-size/)`.

So there's a real chance that they may not pass in a `choiceSize`, and if I did:

```
if (typeof choiceSize !== 'undefined' && choiceImageSizes.includes(choiceSize)) { 
  // ... 
} else {
  throw Error()
}
```

The first condition would fail, and an invalid error would be thrown.

jeremenichelli commented on src/teams.js:
```diff
@@ -1,58 +1,52 @@
-  if (!isMemberPropValid(members)) {
-    throw `No member provided`
+
+  get() {
```
Is this public? If we are changing not internal but API method names the commits should be tagged as breaking changes.

jeremenichelli replied:
I have the feeling that just the word _get_ isn't explicit enough, we should have something clearer around what we are returning here in my opinion.

MichaelSolati replied:
@jeremenichelli It is public => https://developer.typeform.com/create/teams/#walkthroughs

While we could offer a more explicit name (`getMine`?) I didn't want to necessarily cause any breaking changes, unless we renamed the function and then use `get` as a pass through to the renamed function for now.

jeremenichelli replied:
Well, if it's public and we change its name it is a breaking change. It's fine, but you need to label the commit as `BREAKING CHANGE` in the footer of the commit message so the version bump is correct. about the name, I think that `getMine` is good.

jeremenichelli commented on src/themes.js:
```diff
@@ -1,82 +1,76 @@
+  create({ background, colors, font, hasTransparentButton, name }  = {}) {
+    // check if required properties are defined
+    if ([name, font, colors].includes(undefined)) {
+      throw new Error(`Please add the required fields`);
```
I don't like much both the code style of this check and the message. If I'm missing more than one, as a developer I won't know what I'm exactly missing.

I would prefer individual checks with more specific messages of what I'm missing in the arguments.

jeremenichelli commented on src/themes.js:
```diff
@@ -1,82 +1,76 @@
-  })
-}
+    if (!FONTS_AVAILABLE.includes(font)) {
+      throw new Error(`It's not a valid font`);
```
I would add the name of the font here to make the error more specific and clearer for the developer.

```js
`${font} is not a valid font`
```

jeremenichelli commented on src/themes.js:
```diff
@@ -1,82 +1,76 @@
+    }
+
+    if (!FONTS_AVAILABLE.includes(font)) {
+      throw new Error(`It's not a valid font`);
```
Same as above

jeremenichelli commented on src/typeform.js:
```diff
@@ -25,4 +25,4 @@ export const createClient = (args = {}) => {
 }
 
-export default createClient
+export default createClient;
```
missing EOL character here.

jeremenichelli commented on src/utils.js:
```diff
@@ -1,13 +1,10 @@
-  }
+export const buildUrlWithParams = (url, params = {}) => {
+  const queryParams = Object.keys(params)
+    .filter((k) => params[k] !== undefined && params[k] !== null)
```
I would change _k_ for some more meaningful name. Code readability helps maintainability for people that haven't touch the code..

jeremenichelli commented:
Hey @MichaelSolati I left a bunch of comments, feel free to comment on the ones that aren't clear or need further explanation.

I have an extra comment which is code style. I see you added a lot of semicolon characters, the code style across the whole company is to follow the _standard_ code style (https://standardjs.com/) and I think that even if it's not of the maintainers preference we should stick to it. I personally don't like/use standard but I use it because is a crossteam decision.

About the future TypeScript migration, I have the same opinion. We don't use it, so I wouldn't go with it, with JSDoc plus `// @tscheck` comment you can achieve type intellisense. If not, you are putting an entry barrier for internal developers to collaborate. In addition, I would like to see how many TypeScript developers user we have from all the JavaScript spectrum of developers.

Let me know what you think. Cheers!

mbondyra commented on src/forms.js:
```diff
@@ -1,74 +1,78 @@
+      data
+    });
+  }
+}
```
maybe it would be useful to add standard to the project to avoid code style differences (semicolons, empty lines at the end)

MichaelSolati replied:
I added a lint step in the build, so it will fail if it's not standard compliant, as well as an eslint file that extends standard so you can see while you're coding the mistakes you've made

MichaelSolati commented:
> Hey @MichaelSolati I left a bunch of comments, feel free to comment on the ones that aren't clear or need further explanation.
> 
> I have an extra comment which is code style. I see you added a lot of semicolon characters, the code style across the whole company is to follow the _standard_ code style (https://standardjs.com/) and I think that even if it's not of the maintainers preference we should stick to it. I personally don't like/use standard but I use it because is a crossteam decision.
> 
> About the future TypeScript migration, I have the same opinion. We don't use it, so I wouldn't go with it, with JSDoc plus `// @tscheck` comment you can achieve type intellisense. If not, you are putting an entry barrier for internal developers to collaborate. In addition, I would like to see how many TypeScript developers user we have from all the JavaScript spectrum of developers.
> 
> Let me know what you think. Cheers!

@jeremenichelli o I updated the code to be valid with standard (I cried every step of the way 😝), I added a lint step in the travis build to verify it's written the right way and I also added an eslint that just extends the rules of standard. I didn't think/hope that would be an issue, but it allowed for me to see in real time what my errors were.

As far as the TypeScript is concerned, that's an issue for another day. I see a MASSIVE win when it comes to having type definitions/interfaces for all of the forms/blocks. But I can worry about that later.

jeremenichelli commented:
I answered to your comments @MichaelSolati, about the linting, it is not an issue at all, but it increases the diff constantly if everybody who makes a change adds their own preference changes, and makes it harder to review.

About TypeScript, I agree with the benefits, not saying it is useful, but we don't use it across the company which means that we are increasing the entry barrier if another developer wants/needs to change something in the library. But again, discussion for another time.

joshuaslate commented on src/images.js:
```diff
@@ -1,81 +1,79 @@
+      if (['default', 'thumbnail', 'mobile', 'tablet'].includes(backgroundSize)) {
+        requestQuery.url += `/background/${backgroundSize}`
+      } else {
+        throw new Error(`Image background size doesn't exists`)
```
It looks like there's a little typo here-- exists should be singular.
